File: NestCash/backend/app/models/limit.py
# app/models/limit.py
from __future__ import annotations
from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field, validator
from typing import Optional, Literal
from datetime import datetime, timedelta
from enum import Enum

class LimitType(str, Enum):
    SPENDING = "spending"  # Általános kiadási limit
    CATEGORY = "category"  # Kategória-specifikus limit
    ACCOUNT = "account"    # Számla-specifikus limit

class LimitPeriod(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    YEARLY = "yearly"

class Limit(Document):
    user_id: PydanticObjectId
    name: str = Field(..., description="Limit neve (pl. 'Havi élelmiszer budget')")
    type: LimitType = Field(..., description="Limit típusa")
    amount: float = Field(..., gt=0, description="Limit összeg (mindig pozitív)")
    period: LimitPeriod = Field(..., description="Limit időszaka")
    currency: str = Field(default="HUF", description="Deviza")
    
    # Opcionális mezők típusfüggően
    category: Optional[str] = Field(None, description="Kategória neve (category típusnál kötelező)")
    main_account: Optional[str] = Field(None, description="Főszámla (account típusnál)")
    sub_account_name: Optional[str] = Field(None, description="Alszámla neve (account típusnál)")
    
    # Állapot és időzítés
    is_active: bool = Field(default=True, description="Aktív-e a limit")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Értesítési beállítások
    notification_threshold: Optional[float] = Field(None, ge=0, le=100, description="Értesítési küszöb %-ban")
    notify_on_exceed: bool = Field(default=True, description="Értesítés túllépéskor")
    
    @validator('category')
    def validate_category_for_type(cls, v, values):
        """Kategória típusnál kötelező a kategória megadása"""
        if values.get('type') == LimitType.CATEGORY and not v:
            raise ValueError('Kategória típusú limitnél kötelező a kategória megadása')
        return v
    
    @validator('main_account')
    def validate_account_for_type(cls, v, values):
        """Account típusnál kötelező a főszámla megadása"""
        if values.get('type') == LimitType.ACCOUNT and not v:
            raise ValueError('Számla típusú limitnél kötelező a főszámla megadása')
        if v and v not in ['likvid', 'befektetes', 'megtakaritas']:
            raise ValueError('Érvénytelen főszámla típus')
        return v
    
    @validator('notification_threshold')
    def validate_threshold(cls, v):
        """Értesítési küszöb validálása"""
        if v is not None and (v < 0 or v > 100):
            raise ValueError('Az értesítési küszöbnek 0 és 100 között kell lennie')
        return v
    
    async def save(self, **kwargs):
        """Mentés előtt updated_at frissítése"""
        self.updated_at = datetime.utcnow()
        return await super().save(**kwargs)
    
    def get_period_start(self, reference_date: datetime = None) -> datetime:
        """Aktuális időszak kezdetének kiszámítása"""
        if reference_date is None:
            reference_date = datetime.utcnow()
            
        if self.period == LimitPeriod.DAILY:
            return reference_date.replace(hour=0, minute=0, second=0, microsecond=0)
        elif self.period == LimitPeriod.WEEKLY:
            # Hét kezdete (hétfő)
            days_since_monday = reference_date.weekday()
            week_start = reference_date - timedelta(days=days_since_monday)
            return week_start.replace(hour=0, minute=0, second=0, microsecond=0)
        elif self.period == LimitPeriod.MONTHLY:
            return reference_date.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        elif self.period == LimitPeriod.YEARLY:
            return reference_date.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
    
    def get_period_end(self, reference_date: datetime = None) -> datetime:
        """Aktuális időszak végének kiszámítása"""
        from datetime import timedelta
        import calendar
        
        if reference_date is None:
            reference_date = datetime.utcnow()
            
        if self.period == LimitPeriod.DAILY:
            return reference_date.replace(hour=23, minute=59, second=59, microsecond=999999)
        elif self.period == LimitPeriod.WEEKLY:
            # Hét vége (vasárnap)
            days_since_monday = reference_date.weekday()
            week_end = reference_date + timedelta(days=(6 - days_since_monday))
            return week_end.replace(hour=23, minute=59, second=59, microsecond=999999)
        elif self.period == LimitPeriod.MONTHLY:
            # Hónap utolsó napja
            last_day = calendar.monthrange(reference_date.year, reference_date.month)[1]
            return reference_date.replace(day=last_day, hour=23, minute=59, second=59, microsecond=999999)
        elif self.period == LimitPeriod.YEARLY:
            return reference_date.replace(month=12, day=31, hour=23, minute=59, second=59, microsecond=999999)
    
    class Settings:
        name = "limits"
        indexes = [
            "user_id",
            "type",
            "is_active",
            [("user_id", 1), ("is_active", 1)],  # Aktív limitek lekérdezéséhez
            [("user_id", 1), ("type", 1)],       # Típus szerinti lekérdezésekhez
            [("user_id", 1), ("category", 1)],   # Kategória specifikus limitek
        ]

File: NestCash/frontend/lib/models/limit.dart
// lib/models/limit.dart

enum LimitType {
  spending('spending', 'Általános kiadási limit'),
  category('category', 'Kategória specifikus limit'),
  account('account', 'Számla specifikus limit');

  const LimitType(this.value, this.displayName);
  final String value;
  final String displayName;

  static LimitType fromString(String value) {
    return LimitType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => LimitType.spending,
    );
  }
}

enum LimitPeriod {
  daily('daily', 'Napi'),
  weekly('weekly', 'Heti'),
  monthly('monthly', 'Havi'),
  yearly('yearly', 'Éves');

  const LimitPeriod(this.value, this.displayName);
  final String value;
  final String displayName;

  static LimitPeriod fromString(String value) {
    return LimitPeriod.values.firstWhere(
      (period) => period.value == value,
      orElse: () => LimitPeriod.monthly,
    );
  }
}

class Limit {
  final String id;
  final String userId;
  final String name;
  final LimitType type;
  final double amount;
  final LimitPeriod period;
  final String currency;
  final String? category;
  final String? mainAccount;
  final String? subAccountName;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final double? notificationThreshold;
  final bool notifyOnExceed;
  
  // Számított mezők
  final double? currentSpending;
  final double? remainingAmount;
  final double? usagePercentage;
  final DateTime? periodStart;
  final DateTime? periodEnd;

  const Limit({
    required this.id,
    required this.userId,
    required this.name,
    required this.type,
    required this.amount,
    required this.period,
    this.currency = 'HUF',
    this.category,
    this.mainAccount,
    this.subAccountName,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
    this.notificationThreshold,
    this.notifyOnExceed = true,
    this.currentSpending,
    this.remainingAmount,
    this.usagePercentage,
    this.periodStart,
    this.periodEnd,
  });

  factory Limit.fromJson(Map<String, dynamic> json) {
    return Limit(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      name: json['name'] ?? '',
      type: LimitType.fromString(json['type'] ?? 'spending'),
      amount: (json['amount'] ?? 0).toDouble(),
      period: LimitPeriod.fromString(json['period'] ?? 'monthly'),
      currency: json['currency'] ?? 'HUF',
      category: json['category'],
      mainAccount: json['main_account'],
      subAccountName: json['sub_account_name'],
      isActive: json['is_active'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      notificationThreshold: json['notification_threshold']?.toDouble(),
      notifyOnExceed: json['notify_on_exceed'] ?? true,
      currentSpending: json['current_spending']?.toDouble(),
      remainingAmount: json['remaining_amount']?.toDouble(),
      usagePercentage: json['usage_percentage']?.toDouble(),
      periodStart: json['period_start'] != null 
          ? DateTime.parse(json['period_start']) 
          : null,
      periodEnd: json['period_end'] != null 
          ? DateTime.parse(json['period_end']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'type': type.value,
      'amount': amount,
      'period': period.value,
      'currency': currency,
      if (category != null) 'category': category,
      if (mainAccount != null) 'main_account': mainAccount,
      if (subAccountName != null) 'sub_account_name': subAccountName,
      if (notificationThreshold != null) 'notification_threshold': notificationThreshold,
      'notify_on_exceed': notifyOnExceed,
    };
  }

  Limit copyWith({
    String? id,
    String? userId,
    String? name,
    LimitType? type,
    double? amount,
    LimitPeriod? period,
    String? currency,
    String? category,
    String? mainAccount,
    String? subAccountName,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    double? notificationThreshold,
    bool? notifyOnExceed,
    double? currentSpending,
    double? remainingAmount,
    double? usagePercentage,
    DateTime? periodStart,
    DateTime? periodEnd,
  }) {
    return Limit(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      type: type ?? this.type,
      amount: amount ?? this.amount,
      period: period ?? this.period,
      currency: currency ?? this.currency,
      category: category ?? this.category,
      mainAccount: mainAccount ?? this.mainAccount,
      subAccountName: subAccountName ?? this.subAccountName,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      notificationThreshold: notificationThreshold ?? this.notificationThreshold,
      notifyOnExceed: notifyOnExceed ?? this.notifyOnExceed,
      currentSpending: currentSpending ?? this.currentSpending,
      remainingAmount: remainingAmount ?? this.remainingAmount,
      usagePercentage: usagePercentage ?? this.usagePercentage,
      periodStart: periodStart ?? this.periodStart,
      periodEnd: periodEnd ?? this.periodEnd,
    );
  }

  bool get isExceeded => (currentSpending ?? 0) > amount;
  
  bool get isNearLimit {
    if (notificationThreshold == null || currentSpending == null) return false;
    return (usagePercentage ?? 0) >= notificationThreshold!;
  }

  String get statusText {
    if (isExceeded) return 'Túllépve';
    if (isNearLimit) return 'Közel a limithez';
    return 'Rendben';
  }

  String get typeDisplayName {
    switch (type) {
      case LimitType.spending:
        return 'Általános';
      case LimitType.category:
        return 'Kategória: ${category ?? "Nincs megadva"}';
      case LimitType.account:
        return 'Számla: ${mainAccount ?? ""}${subAccountName != null ? " - $subAccountName" : ""}';
    }
  }
}

class LimitCheckResult {
  final bool isAllowed;
  final List<String> exceededLimits;
  final List<String> warnings;
  final String? message;

  const LimitCheckResult({
    required this.isAllowed,
    this.exceededLimits = const [],
    this.warnings = const [],
    this.message,
  });

  factory LimitCheckResult.fromJson(Map<String, dynamic> json) {
    return LimitCheckResult(
      isAllowed: json['is_allowed'] ?? true,
      exceededLimits: List<String>.from(json['exceeded_limits'] ?? []),
      warnings: List<String>.from(json['warnings'] ?? []),
      message: json['message'],
    );
  }
}

class LimitStatus {
  final int totalLimits;
  final int activeLimits;
  final int exceededLimits;
  final int warningLimits;

  const LimitStatus({
    required this.totalLimits,
    required this.activeLimits,
    required this.exceededLimits,
    required this.warningLimits,
  });

  factory LimitStatus.fromJson(Map<String, dynamic> json) {
    return LimitStatus(
      totalLimits: json['total_limits'] ?? 0,
      activeLimits: json['active_limits'] ?? 0,
      exceededLimits: json['exceeded_limits'] ?? 0,
      warningLimits: json['warning_limits'] ?? 0,
    );
  }
}